### ISSUE TITLE
Suggestion: Reopen static and instance side of classes

### ISSUE THREAD

**Issue by rbuckton** [Microsoft/TypeScript#2957]
# Summary
To support both the semantics of subclassing built-ins in ES6 and still allow authors to augment built-ins, we need a mechanism to reopen the static and instance sides of a class.

# Current state
Today we can re-open interfaces, allowing authors to augment built-ins (for example, to support polyfills):
```ts
// in lib.d.ts
interface Array<T> { /*...*/ }
interface ArrayConstructor { /*...*/ }
declare var Array: ArrayConstructor;

// in polyfill.ts
interface Array<T> {
  includes(value: T): Boolean;
}
interface ArrayConstructor {
  of<T>(...items: T[]): Array<T>;
}

Array.prototype.includes = function (value: any) { return this.indexOf(value) != -1; }
Array.of = function<T> (...items: T[]) { return items; }
```

We can also re-open the static side of a class, in a limited fashion:
```ts
// initial declaration
class MyClass {
}

// re-open
module MyClass {
  export var staticProperty = 1;
}
```

There are several issues with these approaches:

* You cannot use type defined by the `var`/`interface` pattern in the `extends` clause of a class in TypeScript, meaning that "classes" defined using this pattern cannot be subclassed in ES6, which is an issue for built-ins.
* While you can re-open the static side of a class using `module`, you can only use non-keyword identifiers for property names. So you could not, for example, add a `[Symbol.species]` property to the class, or use decorators on these members.
* There is no way to re-open the instance side of a class.

# Proposal
I propose we add a new syntactic modifier for the `class` declaration that would indicate we are re-opening an existing class. For this example I am using the keyword `partial`, although the semantics here differ significantly than the same-named capability in C#:

```ts
// in lib.d.ts
declare class Array<T> {
}

// in polyfill.ts
partial class Array<T> {
  static of<T>(...items: T[]) { return items; }
  includes(value: T): boolean { return this.indexOf(value) != -1; }
}
```

## Rules

* A `partial` class declaration must be preceded by a non-`partial` class declaration in the same lexical scope.
* A `partial` class declaration must have the same module visibility as the preceding non-`partial` class declaration.
* A `partial` class declaration must have the same generic type parameters (including constraints) as the non-`partial` class declaration.
* A `partial` class declaration cannot have an `extends` clause, but may have an `implements` clause.
* A `partial` class declaration cannot have a `constructor` member.
* A `partial` class declaration cannot have members with the same name as existing members on a class.
  * Exception: ambient partial class declaration members can merge with other ambient partial class declaration members if they are compatible overloads, similar to interfaces.

# Out of scope
This proposal does not cover the case where built-ins "classes" can often also be called as functions. This case will be covered in a separate proposal.

**Comment by danquirk** [comment]
Note comments in #563

**Comment by RyanCavanaugh** [comment]
Also worth linking #9 since it covers basically the same use cases

**Comment by kode4food** [comment]
yes, please!

**Comment by RyanCavanaugh** [comment]
Need to write up the status of this one

**Comment by MortenHoustonLudvigsen** [comment]
I like this suggestion, but would like to see it extended slightly.

Say I have a class generated from a C# class. I would like to be able to apply decorators to properties of the class without having to change the generated code.

I could see this working, if I am allowed to add an existing property to the partial class with the same type or a more general compatible type (so `any` would work in all cases). The property would *not* change type.

Example:

```typescript
// in Person.ts (a generated file from somewhere)
export class Person {
    @key
    id: number;
    name: string;
}

// in PersonExtensions.ts
import { Person } from './Person';

partial class Person {
    @required
    name: any;
}

```

**Comment by manast** [comment]
+1

**Comment by qwertie** [comment]
In JS you do this by extending the prototype. While a friendly syntax would be welcome, supporting the traditional style would be nice for those that need it. TS is rumored to be a superset of JS after all 🙂
~~~js
Klass.prototype.foo = function() { console.log("foo!"); }
new Klass().foo();
~~~

**Comment by manast** [comment]
@qwertie aren't you just stating the obvious?

**Comment by JohnArcher** [comment]
Is there any progression on this topic? I have the same problem as @trusktr .

**Comment by saschanaz** [comment]
Random bikeshed as the original `partial class` proposal violates the design goal:

```ts
declare class Foo {};
interface class Foo {
  staticMethod(): boolean;
};

let result = Foo.staticMethod(); // boolean
```

**Comment by saschanaz** [comment]
Actually this already works:

```ts
declare class Foo {
    static abc(): void ;
}

declare namespace Foo {
    export function bcd(): void;
}

Foo.abc();
Foo.bcd();
```

So maybe this should just be closed.

**Comment by jcalz** [comment]
I think this issue is also tracking adding new overloads to the construct signature, which can't be done with `namespace`... (although  proposal in the suggestion OP does not seem to allow that either).  So this should probably stay open.

**Comment by Jamesernator** [comment]
This idea would also be useful for extending builtins with private fields (to emulate internal slots) to prevent structural typing in builtins. e.g.:

```ts
declare partial class ArrayBuffer {
    #arrayBufferData: unknown;
}

const arrayBufferLike = { 
    byteLength: 200,
    slice: (begin: number, end?: number) => new ArrayBuffer(20),
};
// Would now report an error as this doesn't actually work in practice
const dataView = new DataView(arrayBufferLike);
```

**Comment by Jamesernator** [comment]
So a problem today is not all interfaces even have `BlahConstructor` interfaces to actually extend. For example `AbortSignal` is defined inline like:

```ts
declare var AbortSignal: {
    prototype: AbortSignal;
    new(): AbortSignal;
    /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/AbortSignal/abort_static) */
    abort(reason?: any): AbortSignal;
    /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/AbortSignal/timeout_static) */
    timeout(milliseconds: number): AbortSignal;
};
```

which makes it impossible to add [missing methods](https://github.com/microsoft/TypeScript/issues/58026) in declarations as there is no `var` merging:

```ts
// Subsequent variable declarations must have the same type.  Variable 'AbortSignal' must be of type '{ new (): AbortSignal; prototype: AbortSignal; abort(reason?: any): AbortSignal; timeout(milliseconds: number): AbortSignal; }', but here has type '{ any(signals: readonly AbortSignal[]): AbortSignal; }'.ts(2403)
lib.dom.d.ts(2337, 13): 'AbortSignal' was also declared here.
declare global {
    var AbortSignal: {
        any(signals: ReadonlyArray<AbortSignal>): AbortSignal;
    };
}
```

If everything were simply defined as `partial class` it'd be trivial:

```ts
declare global {
    partial class AbortSignal {
        static any(signals: ReadonlyArray<AbortSignal>): AbortSignal;
    }
}
```

**Comment by alii** [comment]
Is there a future where this behaviour can be implemented in ambient declarations only? Or at least, allowing for static/constructor overloads in ambient declarations

I'd definitely be happy to give a shot at implementing this

This would unblock the PR https://github.com/microsoft/TypeScript-DOM-lib-generator/pull/1941, which won't merge at the moment because the dom lib generator eventually plans to migrate to class declarations, and the current state of the PR uses the interface/var pattern which would block migration to classes.
